### PR TITLE
ALM External Id Added on RunSet Level

### PR DESCRIPTION
**Pull request by prashelke**
Thank you for your contribution.
Before submitting this PR, please make sure:

- [x] PR description and commit message should describe the changes done in this PR
- [x] Verify the PR is pointing to correct branch i.e. Release or Beta branch if the code fix is for specific release , else point it to master
- [x] Latest Code from master or specific release branch is merged to your branch
- [x] No unwanted\commented\junk code is included
- [x] No new warning upon build solution
- [x] Code Summary\Comments are added to my code which explains what my code is doing
- [ ] Existing unit test cases are passed
- [ ] New Unit tests are added for your development
- [ ] Sanity Tests are successfully executed for New and Existing Functionality
- [x] Verify that changes are compatible with all relevant browsers and platforms.
- [x] After creating pull request there should not be any conflicts
- [ ] Resolve all Codacy comments
- [ ] Builds and checks are passed before PR is sent for review
- [ ] Resolve code review comments
- [ ] Update the Help Library document to match any feature changes

## Summary by CodeRabbit

- **New Features**
	- Added a checkbox for searching ALM entities by name in the export configuration page.
	- Added fields for specifying test cases export folders.
	- Added checkboxes for attaching activities group reports to test cases and using user variables in test case run instance names.

- **Enhancements**
	- Improved visibility and user interaction with the "Search Child Entity By Name" option in the export results configuration.

- **Refactor**
	- Updated project references to ensure better compatibility and performance.

- **Documentation**
	- Corrected comment headers and added new properties documentation for better code clarity and maintenance.

### CODE REVIEW

File: Ginger/Ginger/RunSetPageLib/NewRunSetPage.xaml.cs [hunk 1]
```diff
@@ -903,6 +903,7 @@ void InitRunSetInfoSection()
 
             UpdateDescription();
             xRunDescritpion.Init(mContext, mRunSetConfig, nameof(RunSetConfig.RunDescription));
+            xExternalId.Init(mContext, mRunSetConfig, nameof(RunSetConfig.ExternalID));
             if (mSolutionCategoriesPage == null)
             {
                 mSolutionCategoriesPage = new SolutionCategoriesPage();
```
[the remark this answers is not in the document]
Comment (prashelke) [reply Ginger/Ginger/RunSetPageLib/NewRunSetPage.xaml.cs:909]: updated

File: Ginger/Ginger/Run/RunSetActions/ExportResultsToALMConfigPage.xaml.cs [hunk 2]
```diff
@@ -315,5 +319,15 @@
                 }
             }
         }
+
+        private void SearchChildEntityByName_Unchecked(object sender, RoutedEventArgs e)
+        {
+
+        }
+
+        private void SearchChildEntityByName_Checked(object sender, RoutedEventArgs e)
```
[the remark this answers is not in the document]
Comment (prashelke) [reply Ginger/Ginger/Run/RunSetActions/ExportResultsToALMConfigPage.xaml.cs:328]: updated

File: Ginger/GingerCoreNET/Run/RunSetActions/RunSetActionPublishToQCOperations.cs [hunk 3]
```diff
@@ -187,5 +202,46 @@
...
+                            bFlow.ExternalID = activitiesGroup.ExternalID;
+                        }
+                    }
+                }
+
+            }
+            catch (Exception ex)
+            {
+
+            }
```
[the remark this answers is not in the document]
Comment (prashelke) [reply Ginger/GingerCoreNET/Run/RunSetActions/RunSetActionPublishToQCOperations.cs:244]: updated

File: Ginger/GingerCoreNET/RunLib/DynamicExecutionLib/DynamicExecutionManager.cs [hunk 4]
```diff
@@ -852,6 +852,7 @@ private static GingerExecConfig GetGingerExecConfigurationObject(Solution soluti
                     operationConfigPublishToALM.RunAt = (OperationExecConfigBase.eOperationRunAt?)publishToQCAction.RunAt;
                     operationConfigPublishToALM.AlmTestSetLevel = (AlmPublishOperationExecConfig.eAlmTestSetLevel?)publishToQCAction.ALMTestSetLevel;
                     operationConfigPublishToALM.ExportType = (AlmPublishOperationExecConfig.eExportType?)Enum.Parse(typeof(AlmPublishOperationExecConfig.eExportType), publishToQCAction.ExportType.ToString());
+                    operationConfigPublishToALM.SearchChildALMEntityByName = publishToQCAction.SearchEntityByName;
```
[the remark this answers is not in the document]
Comment (prashelke) [reply Ginger/GingerCoreNET/RunLib/DynamicExecutionLib/DynamicExecutionManager.cs:855]: updated

File: Ginger/Ginger/Run/RunSetActions/ExportResultsToALMConfigPage.xaml [hunk 5]
```diff
@@ -29,6 +29,11 @@
                     <Label Style="{StaticResource @LabelStyle}">Export Type:</Label>
                     <UserControlsLib:UCComboBox x:Name="xExportTypeCbx" ContextMenu="{x:Null}" Margin="10,0,0,0" Width="300" HorizontalAlignment="Left"></UserControlsLib:UCComboBox>
                 </StackPanel>
+                <CheckBox x:Name="SearchChildEntityByName" Style="{StaticResource @CheckBoxStyle}" Margin="5" Checked="SearchChildEntityByName_Checked" Unchecked="SearchChildEntityByName_Unchecked">
```
Comment (Maheshkale447) [inline Ginger/Ginger/Run/RunSetActions/ExportResultsToALMConfigPage.xaml:32]: Change the name to SearchALMEntityByName

File: Ginger/Ginger/Run/RunSetActions/ExportResultsToALMConfigPage.xaml.cs [hunk 6]
```diff
@@ -315,5 +319,15 @@ private void SetTestLevelComboBoxList(RunSetActionBase.eRunAt runAt)
                 }
             }
         }
+
+        private void SearchChildEntityByName_Unchecked(object sender, RoutedEventArgs e)
+        {
```
Comment (Maheshkale447) [inline Ginger/Ginger/Run/RunSetActions/ExportResultsToALMConfigPage.xaml.cs:324]: Do we need this method if its body is empty?

File: Ginger/Ginger/Run/RunSetActions/ExportResultsToALMConfigPage.xaml.cs [hunk 7]
```diff
@@ -315,5 +319,15 @@ private void SetTestLevelComboBoxList(RunSetActionBase.eRunAt runAt)
...
+
+        }
+
+        private void SearchChildEntityByName_Checked(object sender, RoutedEventArgs e)
+        {
```
Comment (Maheshkale447) [inline Ginger/Ginger/Run/RunSetActions/ExportResultsToALMConfigPage.xaml.cs:329]: Do we need this method if its body is empty?

File: Ginger/GingerCore/GingerCore.csproj [hunk 8]
```diff
@@ -62,6 +62,9 @@
     <Reference Include="ALM_CommonStd">
       <HintPath>DLLs\ALM_CommonStd.dll</HintPath>
     </Reference>
+    <Reference Include="Ginger.ExecuterService.Contracts">
```
Comment (Maheshkale447) [inline Ginger/GingerCore/GingerCore.csproj:65]: Open PR for Executor Service Contracts project and use the nugget

File: Ginger/GingerCoreCommon/Repository/BusinessFlowLib/BusinessFlow.cs [hunk 9]
```diff
@@ -1922,5 +1922,8 @@ public void CalculateExternalId(IValueExpression ve)
                 ExternalIdCalCulated = ve.ValueCalculated;
             }
         }
+        public string ALMTestSetLevel { get; set; }
```
Comment (Maheshkale447) [inline Ginger/GingerCoreCommon/Repository/BusinessFlowLib/BusinessFlow.cs:1925]: Why we need to maintain this properties here?

File: Ginger/GingerCoreCommon/Repository/BusinessFlowLib/ActivitiesGroup.cs [hunk 10]
```diff
@@ -521,5 +521,7 @@ public void CalculateExternalId(IValueExpression ve)
             }
         }
 
+        public Guid BusinessFlowGuid { get; set; }
```
Comment (Maheshkale447) [inline Ginger/GingerCoreCommon/Repository/BusinessFlowLib/ActivitiesGroup.cs:524]: BusinessflowGuid can be fetched from parent entity. Is there any specific reason that we need to maintain it again here?

File: Ginger/GingerCoreCommon/Run/RunSetActions/RunSetActionPublishToQC.cs [hunk 11]
```diff
@@ -185,5 +185,11 @@ public override eALMTestSetLevel GetAlMTestSetLevel()
         {
             return ALMTestSetLevel;
         }
+
+        private bool mSearchEntityByName;
```
Comment (Maheshkale447) [inline Ginger/GingerCoreCommon/Run/RunSetActions/RunSetActionPublishToQC.cs:189]: Change it to SearchALmEntityByName

File: Ginger/GingerCoreNET/Run/RunSetActions/RunSetActionPublishToQCOperations.cs [hunk 12]
```diff
@@ -163,7 +177,8 @@ private BusinessFlow ConvertRunSetToBF(IReportInfo reportInfo)
                         ActivitiesGroup virtualAG = new ActivitiesGroup();
                         virtualAG.Name = runSetBF.Name;
                         virtualAG.Description = runSetBF.Description;
-                        
+                        virtualAG.ExternalID = !string.IsNullOrEmpty(runSetBF.ExternalID) ? runSetBF.ExternalID : string.Empty;
+                        virtualAG.BusinessFlowGuid = runSetBF.Guid;
```
Comment (Maheshkale447) [inline Ginger/GingerCoreNET/Run/RunSetActions/RunSetActionPublishToQCOperations.cs:181]: parent Guid can be used here. add the comment what is being stored in it

File: Ginger/GingerCoreNET/Run/RunSetActions/RunSetActionPublishToQCOperations.cs [hunk 13]
```diff
@@ -187,5 +202,46 @@ private BusinessFlow ConvertRunSetToBF(IReportInfo reportInfo)
...
+
+                foreach (GingerRunner runSetrunner in runSetExec.Runners)
+                {
+                    // if executor is null when run if from file
+                    if (runSetrunner.Executor is null)
+                    {
+                        runSetrunner.Executor = new GingerExecutionEngine(runSetrunner);
+                    }
+
+                    List<BusinessFlow> Bflist = WorkSpace.Instance.SolutionRepository.GetAllRepositoryItems<BusinessFlow>().ToList();
```
Comment (Maheshkale447) [inline Ginger/GingerCoreNET/Run/RunSetActions/RunSetActionPublishToQCOperations.cs:228]: remove .ToList()

**Review by Maheshkale447 — CHANGES_REQUESTED** [review]

File: Ginger/Ginger/Run/RunSetActions/ExportResultsToALMConfigPage.xaml.cs [hunk 14]
```diff
@@ -315,5 +319,15 @@
                 }
             }
         }
+
+        private void SearchChildEntityByName_Unchecked(object sender, RoutedEventArgs e)
```
[the remark this answers is not in the document]
Comment (prashelke) [reply Ginger/Ginger/Run/RunSetActions/ExportResultsToALMConfigPage.xaml.cs]: Removed

File: Ginger/Ginger/Run/RunSetActions/ExportResultsToALMConfigPage.xaml.cs (hunk 6 above)
Comment (prashelke) [reply Ginger/Ginger/Run/RunSetActions/ExportResultsToALMConfigPage.xaml.cs]: Removed

File: Ginger/GingerCore/GingerCore.csproj (hunk 8 above)
Comment (prashelke) [reply Ginger/GingerCore/GingerCore.csproj:65]: Removed

File: Ginger/GingerCoreCommon/Repository/BusinessFlowLib/ActivitiesGroup.cs (hunk 10 above)
Comment (prashelke) [reply Ginger/GingerCoreCommon/Repository/BusinessFlowLib/ActivitiesGroup.cs]: Removed used existing filed

File: Ginger/GingerCoreCommon/Run/RunSetActions/RunSetActionPublishToQC.cs (hunk 11 above)
Comment (prashelke) [reply Ginger/GingerCoreCommon/Run/RunSetActions/RunSetActionPublishToQC.cs]: done

File: Ginger/GingerCoreNET/Run/RunSetActions/RunSetActionPublishToQCOperations.cs (hunk 12 above)
Comment (prashelke) [reply Ginger/GingerCoreNET/Run/RunSetActions/RunSetActionPublishToQCOperations.cs]: done

File: Ginger/GingerCoreNET/RunLib/DynamicExecutionLib/DynamicExecutionManager.cs [hunk 15]
```diff
@@ -852,6 +852,7 @@
                     operationConfigPublishToALM.RunAt = (OperationExecConfigBase.eOperationRunAt?)publishToQCAction.RunAt;
                     operationConfigPublishToALM.AlmTestSetLevel = (AlmPublishOperationExecConfig.eAlmTestSetLevel?)publishToQCAction.ALMTestSetLevel;
                     operationConfigPublishToALM.ExportType = (AlmPublishOperationExecConfig.eExportType?)Enum.Parse(typeof(AlmPublishOperationExecConfig.eExportType), publishToQCAction.ExportType.ToString());
+                    operationConfigPublishToALM.SearchALMEntityByName = publishToQCAction.SearchALMEntityByName != null ? publishToQCAction.SearchALMEntityByName : false;
```
[the remark this answers is not in the document]
Comment (prashelke) [reply Ginger/GingerCoreNET/RunLib/DynamicExecutionLib/DynamicExecutionManager.cs:855]: done

File: Ginger/Ginger/Run/RunSetActions/ExportResultsToALMConfigPage.xaml.cs (hunk 7 above)
Comment (prashelke) [reply Ginger/Ginger/Run/RunSetActions/ExportResultsToALMConfigPage.xaml.cs]: removed

File: Ginger/Ginger/Run/RunSetActions/ExportResultsToALMConfigPage.xaml [hunk 16]
```diff
@@ -14,53 +14,58 @@
...
+                <UserControlsLib:UCComboBox x:Name="xExportTypeCbx" ContextMenu="{x:Null}" Margin="145,0,0,0" Width="300" HorizontalAlignment="Left"></UserControlsLib:UCComboBox>
+            </StackPanel>
+            <CheckBox x:Name="SearchALMEntityByName" Style="{StaticResource @CheckBoxStyle}" Margin="5">
+                <CheckBox.Content>
+                    <GingerCore:ucTextDicResource Text="Search ALM Entity By Name"/>
+                </CheckBox.Content>
+            </CheckBox>
+            <StackPanel x:Name="xExportDestinationFolder" Visibility="Collapsed">
+                    <StackPanel Orientation="Horizontal">
+                        <Label Style="{StaticResource @LabelStyle}" Margin="0,10,0,0">Export Test Set Destination Folder:</Label>
```
Comment (Maheshkale447) [inline Ginger/Ginger/Run/RunSetActions/ExportResultsToALMConfigPage.xaml:39]: change it to Test Set Export Folder

File: Ginger/Ginger/Run/RunSetActions/ExportResultsToALMConfigPage.xaml [hunk 17]
```diff
@@ -14,53 +14,58 @@
...
+                        <Actions:UCValueExpression x:Name="xTestSetFolderDestination" Margin="25,10,0,0" Width="300" HorizontalAlignment="Left"/>
+                    </StackPanel>
+                    <StackPanel Orientation="Horizontal">
+                        <Label Style="{StaticResource @LabelStyle}" Margin="0,10,0,0">Export Test Cases Destination Folder:</Label>
```
Comment (Maheshkale447) [inline Ginger/Ginger/Run/RunSetActions/ExportResultsToALMConfigPage.xaml:43]: Change it to Test Cases Export Folder

File: Ginger/GingerCoreNET/Run/RunSetActions/RunSetActionPublishToQCOperations.cs [hunk 18]
```diff
@@ -178,7 +178,7 @@ private BusinessFlow ConvertRunSetToBF(IReportInfo reportInfo)
                         virtualAG.Name = runSetBF.Name;
                         virtualAG.Description = runSetBF.Description;
                         virtualAG.ExternalID = !string.IsNullOrEmpty(runSetBF.ExternalID) ? runSetBF.ExternalID : string.Empty;
-                        virtualAG.BusinessFlowGuid = runSetBF.Guid;
+                        virtualAG.ParentGuid = runSetBF.Guid;//Guid passing in Parent Guid to Update External Id back
```
Comment (Maheshkale447) [inline Ginger/GingerCoreNET/Run/RunSetActions/RunSetActionPublishToQCOperations.cs:181]: Update the comment

**Review by Maheshkale447 — CHANGES_REQUESTED** [review]

File: Ginger/GingerCoreNET/Run/RunSetActions/RunSetActionPublishToQCOperations.cs (hunk 18 above)
Comment (prashelke) [reply Ginger/GingerCoreNET/Run/RunSetActions/RunSetActionPublishToQCOperations.cs:181]: Update

File: Ginger/Ginger/Run/RunSetActions/ExportResultsToALMConfigPage.xaml (hunk 17 above)
Comment (prashelke) [reply Ginger/Ginger/Run/RunSetActions/ExportResultsToALMConfigPage.xaml:43]: Updated

File: Ginger/Ginger/Run/RunSetActions/ExportResultsToALMConfigPage.xaml (hunk 16 above)
Comment (prashelke) [reply Ginger/Ginger/Run/RunSetActions/ExportResultsToALMConfigPage.xaml:39]: Updated

File: Ginger/Ginger/Run/RunSetActions/ExportResultsToALMConfigPage.xaml.cs [hunk 19]
```diff
@@ -160,7 +168,7 @@ private async void xExportToALMBtn_Click(object sender, RoutedEventArgs e)
                 }
             });
             IsProcessing = false;
-            xExportToALMLoadingIcon.Visibility = Visibility.Collapsed;
```
Comment (Maheshkale447) [inline Ginger/Ginger/Run/RunSetActions/ExportResultsToALMConfigPage.xaml.cs:162]: Remove xExportToALMLoadingIcon as it is not being used anymore

File: Ginger/Ginger/Run/RunSetActions/ExportResultsToALMConfigPage.xaml [hunk 20]
```diff
@@ -17,53 +17,63 @@
...
+                    <Label Style="{StaticResource @LabelStyle}">
+                        <Label.Content>
+                             <GingerCore:ucTextDicResource Text="Use User [Variable] TC Run Instance:"/>
+                        </Label.Content>
+                    </Label>
+                    <CheckBox x:Name="UseVariableInTCRunNameCbx" Style="{StaticResource @CheckBoxStyle}" Margin="30,8,0,0" Checked="UseVariableInTCRunNameCbx_Checked" Unchecked="UseVariableInTCRunNameCbx_Unchecked">
+                    </CheckBox>
+            </StackPanel>
+            <StackPanel Margin="5,0,0,15" x:Name="VariableForTCRunNamePanel" Orientation="Horizontal" IsEnabled="False">
+                    <Label Style="{StaticResource @LabelStyle}" Margin="0,10,0,0">TC Run Instance Name :</Label>
```
Comment (Maheshkale447) [inline Ginger/Ginger/Run/RunSetActions/ExportResultsToALMConfigPage.xaml:72]: remove space before ':'

File: Ginger/GingerCore/ALM/RQM/ExportToRQM.cs [hunk 21]
```diff
@@ -783,6 +784,7 @@ public bool ExportBusinessFlowToRQM(BusinessFlow businessFlow, ObservableList<Ex
                             else
                             {
                                 businessFlow.ActivitiesGroups[ActivityGroupCounter].ExternalIdCalculated = ActivityGroupID;
+                                businessFlow.ActivitiesGroups[ActivityGroupCounter].ExternalID = ActivityGroupID;
```
Comment (Maheshkale447) [inline Ginger/GingerCore/ALM/RQM/ExportToRQM.cs:787]: Verify changes for Businessflow level

**Review by Maheshkale447 — CHANGES_REQUESTED** [review]

File: Ginger/Ginger/Run/RunSetActions/ExportResultsToALMConfigPage.xaml (hunk 20 above)
Comment (prashelke) [reply Ginger/Ginger/Run/RunSetActions/ExportResultsToALMConfigPage.xaml]: done

File: Ginger/GingerCore/ALM/RQM/ExportToRQM.cs (hunk 21 above)
Comment (prashelke) [reply Ginger/GingerCore/ALM/RQM/ExportToRQM.cs]: Verified

File: Ginger/Ginger/Run/RunSetActions/ExportResultsToALMConfigPage.xaml.cs (hunk 19 above)
Comment (prashelke) [reply Ginger/Ginger/Run/RunSetActions/ExportResultsToALMConfigPage.xaml.cs:162]: Removed

**Review by Maheshkale447 — APPROVED** [review]

**Pull request merged**